### PR TITLE
Reduce ROM usage due to X509 info

### LINKED ISSUE #2308: Allow omitting mbedtls_x509_crt_info (and mbedtls_x509_crt_verify_info)
### Description
- Type: Enhancement\Feature Request
- Priority: Major

**Justification - why does the library need this feature?**  

On small systems (e.g. Cortex-M4) every KB of code footprint counts, and these functions only seem to be needed in certain debugging/tools contexts. I've noticed that when building with -ffunction-sections, -fdata-sections, and gc-sections, there are a few KB of wasted code space because constant strings are entangled to types in a way that makes it hard for a linker to omit them.

**Suggested enhancement**  

Add two configuration options MBEDTLS_X509_CRT_INFO and MBEDTLS_X509_CRT_VERIFY_INFO to config.h; both enabled by default. When enabled, there is zero change in behavior. When disabled, these APIs and anything that only those APIs use are omitted from the build.

In most cases this is a simple #ifdef wrap of the bits. For the OID descriptors in oid.c, I'm thinking that mbedtls_oid_descriptor_t should not have the string members at all, so that would likely involve adding  OID_DESCRIPTOR and NULL_OID_DESCRIPTOR macros that either include or exclude the two strings. This may mean that builds with and without MBEDTLS_X509_CRT_INFO aren't ABI compatible, but I'm not sure that matters- any system that would favor sharing a library (Linux, Windows) is probably big enough that this optimization wouldn't be of interest.

Notes:
* I have patches for the above approach that are working on my product, although I haven't attempted to create any tests. I'm thinking to fork an existing test and make sure things build when these options are omitted. I'll open a PR once I get those tests created.
* I am, of course, open to improvements for any of this.

### PR DESCRIPTION
**Pull request by pkolbus**
## Description
Allow projects to save some ROM by omitting mbedtls_x509_*_info() and strings only used by these functions. Resolves #2308 

## Status
**READY**

## Requires Backporting
NO - this is an enhancement

## Migrations
NO (I don't think so, but see question below)

## Additional comments
n/a

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] Backported (n/a)

## Steps to test or reproduce

- In small: `scripts/config.pl unset MBEDTLS_X509_INFO; make; make test` (this is included in all.sh; test-ref-configs.pl also exercises several configs with both options unset).
- Compare code size with and without these options set

### CODE REVIEW

**Comment by RonEld** [comment]
As mentioned in #2313 , this is pending CLA as well

**Comment by pkolbus** [comment]
Thanks for the feedback, will rework.

Just for clarification- would you consider mbedtls_x509_crt_verify_info (my second patch) as falling under the same, single configuration option?

**Comment by hanno-arm** [comment]
@pkolbus At first I thought it could give independent value for debugging to have just the verification failure info enabled, but on second thought, a constrained device would rather just export the error code code and have the user look up its meaning manually. Therefore, I'd suggest you also include the removal of the verification info into the scope of the new option.

@mpg What's your view on that?

**Comment by mpg** [comment]
@hanno-arm I agree. We want to support very constrained devices, and we don't want a plethora of config option because combinatorial explosion hurts testing. So we should have only one config option, and it should remove as much as sensible.

**Comment by pkolbus** [comment]
@mpg @hanno-arm, thanks for the great feedback- makes perfect sense. It also looks like mbedtls_debug_print_crt() is basically useless without MBEDTLS_X509_INFO, so I'm planning to guard that whole function.

Should have a rebased and reworked PR tonight (US time); at the moment my PC is working through all.sh.

**Comment by pkolbus** [comment]
The rebase from ba777bf to 9f54fc3 addresses feedback from @hanno-arm 
The rebase from 9f54fc3 to 49955ff updates the test in all.sh for #2325

In my environment, all.sh now stops in "build_default_make_gcc_and_cxx: test: verify header list in cpp_dummy_build.cpp"; I'll look into that this weekend.

**Comment by pkolbus** [comment]
Hi all, sorry for the delay. Finally found some time to rebase and retest. This should be ready to review again.

File: include/mbedtls/oid.h [hunk 1]
```diff
@@ -588,6 +591,7 @@ int mbedtls_oid_get_md_hmac( const mbedtls_asn1_buf *oid, mbedtls_md_type_t *md_
  * \return         0 if successful, or MBEDTLS_ERR_OID_NOT_FOUND
  */
 int mbedtls_oid_get_extended_key_usage( const mbedtls_asn1_buf *oid, const char **desc );
+#endif
```
Comment (hanno-arm) [inline include/mbedtls/oid.h]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

File: include/mbedtls/x509_crl.h [hunk 2]
```diff
@@ -149,6 +150,7 @@ int mbedtls_x509_crl_parse_file( mbedtls_x509_crl *chain, const char *path );
  */
 int mbedtls_x509_crl_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_crl *crl );
+#endif
```
Comment (hanno-arm) [inline include/mbedtls/x509_crl.h]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

File: include/mbedtls/x509_crt.h [hunk 3]
```diff
@@ -362,7 +363,9 @@ int mbedtls_x509_crt_parse_path( mbedtls_x509_crt *chain, const char *path );
  */
 int mbedtls_x509_crt_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_crt *crt );
+#endif
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Minor: These two `#endif ... #if` can be merged.

File: include/mbedtls/x509_crt.h [hunk 4]
```diff
@@ -377,6 +380,7 @@ int mbedtls_x509_crt_info( char *buf, size_t size, const char *prefix,
  */
 int mbedtls_x509_crt_verify_info( char *buf, size_t size, const char *prefix,
                           uint32_t flags );
+#endif
```
Comment (hanno-arm) [inline include/mbedtls/x509_crt.h]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

File: include/mbedtls/x509_csr.h [hunk 5]
```diff
@@ -137,6 +138,7 @@ int mbedtls_x509_csr_parse_file( mbedtls_x509_csr *csr, const char *path );
  */
 int mbedtls_x509_csr_info( char *buf, size_t size, const char *prefix,
                    const mbedtls_x509_csr *csr );
+#endif
```
Comment (hanno-arm) [inline include/mbedtls/x509_csr.h]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

File: library/oid.c [hunk 6]
```diff
@@ -46,6 +46,17 @@
...
+/*
+ * Macro to generate mbedtls_oid_descriptor_t
+ */
+#if defined(MBEDTLS_X509_INFO)
+#define OID_DESCRIPTOR(s, name, description)  { ADD_LEN(s), name, description }
+#define NULL_OID_DESCRIPTOR                   { NULL, 0, NULL, NULL }
+#else
+#define OID_DESCRIPTOR(s, name, description)  { ADD_LEN(s) }
+#define NULL_OID_DESCRIPTOR                   { NULL, 0 }
+#endif
```
Comment (hanno-arm) [inline library/oid.c]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

File: programs/ssl/ssl_client2.c [hunk 7]
```diff
@@ -583,13 +586,19 @@ static int my_verify( void *data, mbedtls_x509_crt *crt,
 
     mbedtls_printf( "\nVerify requested for (Depth %d):\n", depth );
     mbedtls_printf( "%s", buf );
+#else
+    ((void) crt);
+    ((void) depth);
+#endif
```
Comment (hanno-arm) [inline programs/ssl/ssl_client2.c]: Minor: Please add `/* MBEDTLS_X509_INFO */` here.

**Comment by hanno-arm** [comment]
@mpg Do you think you'll be able to find some time to re-review this, too?

**Comment by hanno-arm** [comment]
@pkolbus Do you think you'll have time to finalize this PR, or shall we take over?

**Comment by pkolbus** [comment]
Yes, I can continue working on this. Was waiting a bit for @mpg to respond and lost track of it.

**Comment by pkolbus** [comment]
After rebasing, it looks like some/all of the `all.sh` tests are building with the crypto submodule, which is causing some build errors when `MBEDTLS_X509_INFO` is disabled. I think this design now requires a PR to mbed-crypto too, plus a new config for the `oid.c` and `oid.h` changes since x509 isn’t part of crypto. I can take care of all that.

Does this plan make sense? Would `MBEDTLS_OID_NAMES` be reasonable?

(If the mbedcrypto split is in a high degree of flux right now and it would make more sense to postpone this PR a bit, I’m happy to hold off)

**Comment by Patater** [comment]
Hi @pkolbus,

Yes, the mbedcrypto split does complicate things a bit. It's not ideal right now, but because the previous libmbedcrypto library had oid.c in it, Mbed TLS currently gets that from Mbed Crypto. This is the final shape we want for these libraries, but the current shape is likely what we'll continue with for the next couple of months.

Could you raise any oid changes to Mbed Crypto? After that, make sure this PR includes an update to the submodule, pointing at a version of Mbed Crypto that has your oid changes.

Thanks, and sorry for the inconvenience.

**Comment by hanno-arm** [comment]
Hi @pkolbus,

something I didn't spot earlier: The new compile-time option `MBEDTLS_X509_INFO` breaks the API for users which use a standalone configuration of Mbed TLS, because they'll loose the CRT info functionality unless they modify their configuration to include `MBEDTLS_X509_INFO`.

I'd suggest (ping @Patater) to negate the option to something like `MBEDTLS_X509_REMOVE_INFO` to resolve this.

Kind regards,
Hanno

**Comment by hanno-arm** [comment]
Hi @pkolbus,

As you might be aware, your PR made it into the [baremetal](https://github.com/ARMmbed/mbedtls/blob/baremetal/include/mbedtls/config.h#L2044-L2053) branch, but unfortunately the development version got stuck.

Would you be willing to rebase the PR on top of development and negate the configuration option to `MBEDTLS_X509_REMOVE_INFO`, as explained in my last comment?

If you don't have time for this, I can also do the necessary rework - let me know what works best for you.

Thank you again for your contribution, it played an important in the RAM/ROM reductions that set the baremetal branch apart.

Kind regards,
Hanno

**Comment by hanno-arm** [comment]
Superseded by https://github.com/ARMmbed/mbedtls/pull/3777 which is a rebased and reworked version of this PR.

**Pull request closed without merging**